### ISSUE TITLE
redis keys explosion of sixpack

### ISSUE THREAD

**Issue by SimpleExpress** [seatgeek/sixpack#246]
I deployed sixpack 2.3.1 in production, and started it using supervisor with the configurations:
```
[program:sixpack]
command=gunicorn --access-logfile - -w 8 -k gevent -b 0.0.0.0:5000 sixpack.server:start
autorestart=true
environment=SIXPACK_CONFIG="/etc/sixpack/config.yml"
stdout_logfile=/var/log/sixpack/sixpack-server.log
redirect_stderr=true

[program:sixpack-web]
command=sixpack-web
autorestart=true
environment=SIXPACK_CONFIG="/etc/sixpack/config.yml"
stdout_logfile=/var/log/sixpack/sixpack-web.log
redirect_stderr=true
```

I finished the deployment at last Friday and it worked well. But when I opened the web page today (Monday) I just found the page hanged. I tried to find what have happened and found out that the redis had keys explosion which was used by sixpack:
```
# Keyspace
db15:keys=9628,expires=0,avg_ttl=0
```

some keys I found in redis (by typing keys *)
```
9625) "sxp:p:almark_test UNION ALL SELECT CONCAT(0x6b5350654d70,0x43786d484968),CONCAT(0x6b5350654d70,0x43786d484968),CONCAT(0x6b5350654d70,0x43786d484968),CONCAT(0x6b5350654d70,0x43786d484968)#:_all:2016"
9626) "sxp:p:-1869') UNION ALL SELECT CONCAT(0x4b475367494b,0x5659704c4f43),CONCAT(0x4b475367494b,0x5659704c4f43),CONCAT(0x4b475367494b,0x5659704c4f43),CONCAT(0x4b475367494b,0x5659704c4f43),CONCAT(0x4b475367494b,0x5659704c4f43),CONCAT(0x4b475367494b,0x5659704c4f43),CONCAT(0x4b475367494b,0x5659704c4f43),CONCAT(0x4b475367494b,0x5659704c4f43),CONCAT(0x4b475367494b,0x5659704c4f43),CONCAT(0x4b475367494b,0x5659704c4f43),CONCAT(0x4b475367494b,0x5659704c4f43),CONCAT(0x4b475367494b,0x5659704c4f43),CONCAT(0x4b475367494b,0x5659704c4f43),CONCAT(0x4b475367494b,0x5659704c4f43),CONCAT(0x4b475367494b,0x5659704c4f43)-- :_all:2016-08-06"
9627) "sxp:p:almark_test UNION ALL SELECT CONCAT(0x6b5350654d70,0x43786d484968),CONCAT(0x6b5350654d70,0x43786d484968),CONCAT(0x6b5350654d70,0x43786d484968),CONCAT(0x6b5350654d70,0x43786d484968),CONCAT(0x6b5350654d70,0x43786d484968),CONCAT(0x6b5350654d70,0x43786d484968),CONCAT(0x6b5350654d70,0x43786d484968),CONCAT(0x6b5350654d70,0x43786d484968),CONCAT(0x6b5350654d70,0x43786d484968),CONCAT(0x6b5350654d70,0x43786d484968)#:_all:2016-08"
9628) "sxp:p:-1891 UNION ALL SELECT CONCAT(0x4c5569786f48,0x797865706b71),CONCAT(0x4c5569786f48,0x797865706b71)-- :warm:2016-08"
(4.95s)
```

Since we haven't send any real data to sixpack, it should only contains the data (about 10+ keys) that I used for testing (the almark_test experiment). Can you help to figure out what had happened, and how can I fix this issue. Thanks a lot.

**Comment by SimpleExpress** [comment]
I tried to investigate it and found it was actually an attack. I do appreciate if someone could give some suggestions for the enhancement work.

**Comment by SimpleExpress** [comment]
Tried to debug it, just found the experiment name was not validated. Issue a request like below will insert a new experiment into the redis:
```
http://localhost:5000/participate?experiment=almark_test%27%29+AND+%28SELECT+1293+FROM%28SELECT+COUNT%28*%29%2CCONCAT%280x48645561554b%2C%28SELECT+%28CASE+WHEN+%281293%3D1293%29+THEN+1+ELSE+0+END%29%29%2C0x574248476759%2CFLOOR%28RAND%280%29*2%29%29x+FROM+INFORMATION_SCHEMA.CHARACTER_SETS+GROUP+BY+x%29a%29+AND+%28%27dhv%27%3D%27dhv&alternatives=cool&alternatives=warm&client_id=155051391258
```

I could do some validation for this, but will someone give a more complete solution for the security considerations. Thanks.

**Comment by josegonzalez** [comment]
We don't do name validation because we don't "whitelist" experiments from being created. That was outside of the scope of sixpack, and is actually part of the design document.

If you want to whitelist experiments, I would recommend fronting Sixpack with some sort of proxy mechanism, like `vagrant` or `nginx+lua`/`openresty`, and performing whitelisting at that level.

I'll add a note to our readme to note that this is the case.

**Comment by josegonzalez** [comment]
Closed by b59dfae986dea6e3af51647e3bc68ebb30f5589e
